test/gtest/gpu_nchw_nhwc_transpose.cpp

```diff
@@ -0,0 +1,658 @@
+};
+
+template <>
+struct to_miopen_data_type<uint16_t>
+{
+    static miopenDataType_t get() { return miopenHalf; } // we actually didn't calculate 16bit float
+};
+
+template <>
+struct to_miopen_data_type<uint8_t>
+{
+    static miopenDataType_t get() { return miopenInt8; }
+};
+
+static constexpr int RAND_INTEGER_MAX = 120;
+static constexpr int RAND_INTEGER_MIN = -88;
+
+template <typename T>
+void rand_tensor_integer(tensor<T>& t, int max = RAND_INTEGER_MAX, int min = RAND_INTEGER_MIN)
+{
+    // use integer to random.
+    for(size_t i = 0; i < t.data.size(); i++)
+        t[i] = static_cast<T>(prng::gen_A_to_B(min, max));
+}
+
+template <typename T>
+bool compare_equal(T r1, T r2)
+{
+    return r1 == r2;
+}
+
+template <>
+bool compare_equal<float>(float r1, float r2)
+{
+    return miopen::float_equal(r1, r2);
+}
+
+template <typename T>
+void verify_tensor(tensor<T>& t_tst, const char* _tst, tensor<T>& t_ref, const char* _ref)
+{
+    ASSERT_EQ(t_tst.data.size(), t_ref.data.size()) << " tensor sizes not equal, should not happen";
+    if(t_tst.data.size() != t_ref.data.size())
+        return;
+
+    auto idx = miopen::mismatch_idx(t_tst.data, t_ref.data, compare_equal<T>);
+
+    EXPECT_GE(idx, miopen::range_distance(t_ref))
+        << "diff at:" << idx << ", " << _tst << ":" << t_tst[idx] << ", " << _ref << ":" << t_ref[idx];
+}
+
+struct transpose_dims
+{
+    static constexpr uint32_t image_size_rand_offset {29};
+    static constexpr uint32_t image_size_rand_range {13};
+    static uint32_t get_max_image_size()
+    {
+        return image_size_rand_offset + image_size_rand_range - 1;
+    }
+    static std::vector<uint32_t> get_image_size()
+    {
+        std::vector<uint32_t> v = {1, 9, 14};
+        v.push_back(prng::gen_off_range(image_size_rand_offset, image_size_rand_range));
+        return v;
+    }
+    static std::vector<uint32_t> get_channel_size()
+    {
+        std::vector<uint32_t> v = {3, 11};
+        v.push_back(prng::gen_off_range(19, 7));
+        return v;
+    }
+    static std::vector<uint32_t> get_batch_size()
+    {
+        std::vector<uint32_t> v = {1, 2, 4};
+        v.push_back(prng::gen_off_range(3, 4));
+        return v;
+    }
+};
+
+struct transpose_invoke_param : public miopen::InvokeParams
+{
+    ConstData_t src = nullptr;
+    Data_t dst      = nullptr;
+
+    transpose_invoke_param(ConstData_t src_, Data_t dst_) : src(src_), dst(dst_) {}
+    transpose_invoke_param(miopen::InvokeType type_, ConstData_t src_, Data_t dst_)
+        : InvokeParams{type_}, src(src_), dst(dst_)
+    {
+    }
+
+    Data_t GetWorkspace() const { return nullptr; }
+    std::size_t GetWorkspaceSize() const { return 0; }
+};
+
+} // namespace batched_transpose
+
+using namespace batched_transpose;
+
+template <typename T, class TRANSPOSE_SOL>
+struct LayoutTransposeTest_2D
+    : public ::testing::TestWithParam<
+          std::tuple<uint32_t, uint32_t>>
+{
+protected:
+    miopenHandle_t              handle{};
```

Comment:
use `miopen::Handle` directly and then you don't have to `create` it or `destroy` it.

Reply:
fixed (in layout_transpose.cpp)